### DISCUSSION TITLE
Low memory images?

### DISCUSSION THREAD

**Discussion by yanshay** [slint-ui/slint#4292] [Q&A]
I'm developing for esp32 and short on memory. I need to use images beyond my current available free memory.
I can do with lower color images, even monochrome if required (as long as there's an alpha channel for transparent pixels, or set one color as transparent).
I think that currently images in memory use 4 bytes per pixel, is that correct?
- Is there a way to have images in memory taking less space? I saw there's an internal 1 byte per pixel representation inside Slint for static texture but not for images.
- Could I implement my own image format including required rendering (for example  2 bits for each color and then alpha - all in a single byte per pixel)?
- Any other way to optimize memory management around images?

**Comment by zhangzqs** [comment]
I encountered the same problem as you. I need to use slint, but at the same time I want to display network images, but my ESP32C3 cannot store a complete 240x240 RGB image pixel array in the remaining available memory.
If I don't use slint, I can draw pixels into the framebuffer of the screen line by line without taking up the memory space of the ESP32C3, but if I use Slint, slint will occupy my entire screen, and I can only use the image component to display the picture. , but this will take up a lot of memory space.
My solution is to implement a multiplexer based on embedded-graphics. A screen that physically implements the embedded-graphics interface can be logically abstracted into multiple screens. slint occupies one logical screen. I am using slint. During the process, I can still dynamically switch different logical screens and render pictures on them. When I don't need it, I can still switch back to the logical screen displayed by slint.

https://github.com/zhangzqs/esp-clock-rs/blob/master/slint-app/src/app/photo.rs
https://github.com/zhangzqs/esp-clock-rs/blob/master/libs/embedded-graphics-group/src/lib.rs

**Reply by yanshay** [reply]
Interesting, so you pass touch events to slint only in the area where it owns and translate touch coordinates?

In my case I need slint images, since slint animates them, shows them, etc. just need some ability to render myself my own objects or have lower memory images.

**Comment by ogoffart** [comment]
One thing i was contemplating is to create a slint::Image from a function that renders pixels.
Similar to [BorrowedOpenGLTextureBuilder](https://docs.rs/slint/latest/slint/struct.BorrowedOpenGLTextureBuilder.html#), have a renderer specific way to construct an image from a `Box<dyn Fn(line: usize, range: Range<usize>, buffer: &mut[TargetPixel])>` or something similar. (we would also need something that supports scaling or transforms)

**Reply by yanshay** [reply]
Will this creat an image in memory with full memory footprint? Or will it be rendered as needed on the fly when needed (so relying on develper's custom memory storage and rendering code}?
And will this support all the caching you currently have to speed rendering on low ende devices?

For end developer using the Slint I think it would also be helpful if there were memory efficient formats with optimized rendering performance, so we don't need to develop optimized code for graphic rendering.

**Reply by zhangzqs** [reply]
I think there is no need to do additional optimizations inside slint. Optimization may lead to an increase in memory usage, and the degree of freedom for optimization from the caller's perspective will be reduced.
But slint might be able to provide a decorator like BufRead util to cache some pixels

**Reply by zhangzqs** [reply]
Can this method draw graphics onto an image component through the embedded-graphics api without requiring the full image memory footprint?

**Reply by ogoffart** [reply]
The idea is that the developer using Slint would provide a function that directly write a line into the target buffer.

Internally, you can then represent the image with the format of your choice, then you can use library function or proceduraly generate the pixel.

Instead of a function, this would probably be a trait.

**Reply by ogoffart** [reply]
> Can this method draw graphics onto an image component through the embedded-graphics api without requiring the full image memory footprint?

No, the embedded-graphics draw in a full buffer. it is not compatible with our line by line rendering.

**Comment by zhangzqs** [comment]
> > Can this method draw graphics onto an image component through the embedded-graphics api without requiring the full image memory footprint?
> 
> No, the embedded-graphics draw in a full buffer. it is not compatible with our line by line rendering.

Yes, so maybe we can write another drawing framework that supports lazy rendering instead of having to render to the buffer immediately.
